ProcessLib/ComponentTransport/ComponentTransportFEM.h

```diff
@@ -628,28 +630,49 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
                                velocity_magnitude) *
                           I);
 
-            // matrix assembly
-            local_M.noalias() += w * N.transpose() * retardation_factor *
-                                     porosity * density * N +
-                                 w * N.transpose() * retardation_factor *
-                                     porosity * C_int_pt * drho_dC * N;
+            double const R_times_phi = retardation_factor * porosity;
+            auto const N_t_N = (N.transpose() * N).eval();
+
+            if (_process_data.non_advective_form)
+            {
+                const double drho_dC =
+                    _process_data.fluid_properties->getdValue(
+                        MaterialLib::Fluid::FluidPropertyType::Density,
+                        vars,
+                        MaterialLib::Fluid::PropertyVariableType::C);
+                local_M.noalias() +=
+                    N_t_N * (R_times_phi * C_int_pt * drho_dC * w);
+            }
+            else
+            {
+                local_b.noalias() -=
+                    N.transpose() *
+                    (density * velocity.dot(dNdx * local_C) * w);
```

Comment:
Could be reformulated as: `velocity.dot(dNdx * local_C) * (density * w)`.

Reply:
done